### PR TITLE
POC: Custom pipelines and data-types

### PR DESCRIPTION
**Pull request by owais**
**DO NOT REVIEW TO MERGE**. This is just a POC to be compared with #817. In case we decide to implement anything similar to this, it'll be a new PR.

--------------------

This proposal presents an alternate implementation to support custom data-types to the one proposed in PR #817. 

The proposal is to add a new pipeline data-type/factory type to the collector. A pipeline will be responsible for defining the data type and components that work with the data type. This makes it very simple to add new types of pipelines without having to modify other internal parts of the collector. It also makes it possible to define new pipelines in contrib collector without having to add them to the core. 

The POC re-implements the existing traces pipeline as a pluggable one and also implement a toy hello world pipeline. Current implementation adds a pipeline factory interface that new pipelines need to implement.

```go
type PipelineFactory interface {
	Type() configmodels.Type

	CreateReceiver(context.Context, ReceiverFactoryBase, *zap.Logger, configmodels.Receiver, Consumer) (Receiver, error)
	CreateProcessor(context.Context, ProcessorFactoryBase, *zap.Logger, Consumer, configmodels.Processor) (Processor, error)
	CreateExporter(context.Context, ExporterFactoryBase, *zap.Logger, configmodels.Exporter) (Exporter, error)

	CreateFanOutConnector([]Consumer) Consumer
	CreateCloningFanOutConnector([]Consumer) Consumer
}
```

Both pipelines are hand-coded but can be totally auto-generated. With a fairly simple code generator, the following should be all that is needed to implement a custom pipeline. 

```go
package customPipeline

//go:generate otel-pipeline-generator -type CustomDataType -Name CustomPipeline

type CustomDataType struct {
	Message string
}

func (cd *CustomDataType) Clone() CustomDataType {
	return CustomDataType{Message: cd.Message}
}
```

### CODE REVIEW

File: config/configmodels/configmodels.go [hunk 1]
```diff
@@ -106,19 +106,31 @@ func (dataType DataType) GetString() string {
...
 	Exporters  []string `mapstructure:"exporters"`
 }
 
+// Type gets the pipeline type.
+func (p *Pipeline) Type() Type {
+	return p.TypeVal
+}
+
+// SetType sets the pipeline type.
+func (p *Pipeline) SetType(typeStr Type) {
```
Comment (bogdandrutu) [inline config/configmodels/configmodels.go:130]: Isn't this the data type?

File: exporter/hwexporter/factory.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2019, OpenTelemetry Authors
```
Comment (bogdandrutu) [inline exporter/hwexporter/factory.go:1]: Year?

File: component/consumer.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+package component
+
+type Consumer interface{}
```
Comment (bogdandrutu) [inline component/consumer.go:3]: This should probably be in the Consumer package.

File: component/consumer.go [hunk 4]
```diff
@@ -0,0 +1,3 @@
+package component
```
Comment (bogdandrutu) [inline component/consumer.go:1]: Copyright?

**Comment by tigrannajaryan** [comment]
> Overall not a big fun of this design, since we loose the ability to verify at validation time that the expected data type is correct inside a pipeline between components and this will be determined at runtime (no guarantees for example that a component will always receive Trace data).

@bogdandrutu Is this so? I thought generated code uses strong typing for pipeline data types. Components which implement the particular data type will be use strongly typed consumer functions. I may have missed something.

**Comment by bogdandrutu** [comment]
@tigrannajaryan let me summarize what I see:
* Pros
  * Having a Pipeline factory can help us building some components specific to that pipeline type easier.
  * Allows to support a "default config" per pipeline type, which is nice. This was one of the request from @pavolloffay.
  * Nice consistency - every major component has a factory.
* Cons
  * Some parts are not strongly type, but maybe (not convinced we can do it) we can have a "validateConfig" on the factory to discover some of these problems during creation. For example the "next" component in the factory is `interface{}` which does not guarantee the right type or data format.
  * Unclear how to configure/build a pipeline (maybe two) that can produce metrics from traces as an example. Do we need a custom pipeline factory for that? Maybe was not possible with the old design, but it feels to me that with the new design is impossible.
  * Unclear the difference between input type and type in the pipeline.

I would add initially the PipelineFactory interface in an internal package, it seems to move in the right direction but we may find more problems when we lock carefully at the code and analyze all the possible scenarios.

File: config/configmodels/configmodels.go (hunk 1 above)
Comment (owais) [reply config/configmodels/configmodels.go:130]: It's the type of the pipeline akin to type of a processor/exporter but it does overlap with our concept of datatype. We can model it either way.

**Comment by owais** [comment]
Agree with Bogdan. I wanted to put the idea forward ASAP. For sure we need to think more about specifics but I think it sets us in a good direction. 

> Some parts are not strongly type, but maybe (not convinced we can do it) we can have a "validateConfig" on the factory to discover some of these problems during creation. For example the "next" component in the factory is interface{} which does not guarantee the right type or data format.

By "creation", do you mean at runtime when a pipeline is first generated from config? If so, we already have checks for these. While the collector sees the components as basic consumer, processor, exporter, etc that it can start and stop, the pipelines do type-assertion when they first start so internally they deal with higher level interfaces that process their own data type model. While this is a bit lacking, it is no worse than what we have right now as users can configure a pipeline with wrong type of component today and it would raise same type assertion errors at runtime.

> Unclear how to configure/build a pipeline (maybe two) that can produce metrics from traces as an example. Do we need a custom pipeline factory for that? Maybe was not possible with the old design, but it feels to me that with the new design is impossible.

Is the goal that the same pipeline should be able to handle both or just that it should be possible to emit metrics through a metrics pipeline from a traces (or any other pipeline)?  We have the ability to get references to other components now. 

A basic implementation could look something like where any component can use `host` to get hold of a metrics receiver or the metrics pipeline and emit metrics to that. 

A more sophisticated model could be where the `host` object provides an interface that accepts metrics so any component can export it's own metrics to host. A metrics pipeline could then be configured to export all the metrics received by host. I think something like the following should work I suppose. 

```
receivers:
    otlp/metrics: {}

service:
    hostMetricsReceiver: otlp/metric
    traces: {}
    metrics:
        receivers: [otlp/metrics]
        ...
```

File: go.mod [hunk 5]
```diff
@@ -16,7 +16,7 @@ require (
 	github.com/go-kit/kit v0.9.0
 	github.com/gogo/googleapis v1.3.0 // indirect
 	github.com/golang/groupcache v0.0.0-20200121045136-8c9f03a8e57e
-	github.com/golang/protobuf v1.3.5
+	github.com/golang/protobuf v1.4.0
```
[the remark this answers is not in the document]
Comment (owais) [reply go.mod:19]: Sorry, I just wanted to share the basic idea. Didn't intent for this PR to be merged so it's littered with stuff either should not have come in or should have been removed. 

I can send a new PR or polish this one if/when we decide to go down this path.

**Comment by bogdandrutu** [comment]
@owais I am talking more like if someone wants to calculate the Histogram of spans latency, not the components metrics. 

TraceReceiver -> TraceProcessor -> MetricsProcessor -> MetricsExporter or something along these lines.

**Comment by bogdandrutu** [comment]
It looks like the issue was closed https://github.com/open-telemetry/opentelemetry-collector/issues/944 but still a request that I heard. Not saying we should implement now but something that we should not make impossible to have.

**Comment by owais** [comment]
Agree, we should definitely make sure we don't make this impossible even if we don't implement it now. 

> @owais I am talking more like if someone wants to calculate the Histogram of spans latency, not the components metrics.
> TraceReceiver -> TraceProcessor -> MetricsProcessor -> MetricsExporter or something along these lines.

Yea, that's how I understood it as well. So trace processor would "generate/caclulate" metrics from spans and then export those. I'm saying instead of having a multi-datatype pipeline, we could compose/stich multiple pipelines together. So `TraceProcessor` in your example would calculate metrics and then export them a metrics pipeline instead of the next processor (assuming it'd support metrics). Wouldn't that work?

**Comment by tigrannajaryan** [comment]
@owais any plans to move forward with this or we should table this for now?

**Comment by owais** [comment]
Yes, I'll resume work on this shortly and have a more reviewable version that can be merged.

**Comment by bogdandrutu** [comment]
I will close this for the moment. Feel free to reopen when you have time to work on this.

**Pull request closed without merging**